EIPS/eip-devp2p-snappy.md

```diff
@@ -0,0 +1,192 @@
+
+## Motivation
+Synchronizing the Ethereum main network (block 4,248,000) in Geth using fast sync currently consumes 1.01GB upload and 33.59GB download bandwidth. On the Rinkeby test network (block 852,000) it's 55.89MB upload and 2.51GB download.
+
+However, most of this data (blocks, transactions) are heavily compressable. By enabling compression at the message payload level, we can reduce the previous numbers to 1.01GB upload / 13.46GB download on the main network, and 46.21MB upload / 463.65MB download on the test network.
+
+The motivation behind doing this at the DEVp2p level (opposed to eth for example) is that it would enable compression for all sub-protocols (eth, les, bzz) seamlessly, reducing any complexity those protocols might incur in trying to individually optimize for data traffic.
+
+## Specification
+Bump the advertised DEVp2p version number from `4` to `5`. If during handshake, the remote side advertises support only for version `4`, run the exact same protocol as until now.
+
+If the remote side advertises a DEVp2p version `>= 5`, inject a Snappy compression step right before encrypting the DEVp2p message during sending:
+
+ * A message consists of `{Code, Size, Payload}`
+  * Compress the original payload with Snappy and store it in the same field.
+  * Update the message size to the length of the compressed payload.
+  * Encrypt and send the message as before, oblivious to compression.
+
+Similarly to message sending, when receiving a DEVp2p v5 message from a remote node, insert a Snappy decompression step right after the decrypting the DEVp2p message:
+
+* A message consists of `{Code, Size, Payload}`
+ * Decrypt the message payload as before, oblivious to compression.
+ * Decompress the payload with Snappy and store it in the same field.
+ * Update the message size to the length of the decompressed payload.
+
+Important caveats:
+
+ * The handshake message is **never** compressed, since it is needed to negotiate the common version.
+ * Snappy framing is **not** used, since the DEVp2p protocol already message oriented.
+
+### Avoiding DOS attacks
+
+Currently a DEVp2p message length is limited to 24 bits, amounting to a maximum size of 16MB. With the introduction of Snappy compression, care must be taken not to blindy decompress messages, since they may get significantly larger than 16MB.
+
+However, Snappy is capable of calculating the decompressed size of an input message without inflating it in memory. This can be used to discard any messages which decompress above some threshold. **The proposal is to use the same limit (16MB) as the threshold for decompressed messages.** This retains the same guarantees that the current DEVp2p protocol does, so there won't be surprises in application level protocols.
```

Comment:
Can you extend *Snappy is capable of calculating the decompressed size of an input message without inflating it in memory* with

> The stream starts with the uncompressed length (up to a maximum of 2^32 - 1),
stored as a little-endian varint.

Reply:
Fixed.